lib/collection/src/shards/replica_set/update.rs

```diff
@@ -322,91 +322,92 @@ impl ShardReplicaSet {
-                .await?;
+            // If there are enough successes, deactivate failed replicas
+            // Failed replicas will automatically recover from another replica ensuring consistency
+            if successes.len() >= minimal_success_count {
+                let wait_for_deactivation = self.handle_failed_replicas(
+                    &failures,
+                    &self.replica_state.read(),
+                    update_only_existing,
+                );
 
-                if !shards_disabled {
-                    return Err(CollectionError::service_error(format!(
-                        "Some replica of shard {} failed to apply operation and deactivation \
-                         timed out after {} seconds. Consistency of this update is not guaranteed. Please retry. {failure_error}",
-                        self.shard_id, timeout.as_secs()
-                    )));
+                // Wait for replica failures to be accepted, otherwise return consistency error
+                if wait && wait_for_deactivation && !failures.is_empty() {
+                    // ToDo: allow timeout configuration in API
+                    let timeout = DEFAULT_SHARD_DEACTIVATION_TIMEOUT;
+
+                    let replica_state = self.replica_state.clone();
+                    let peer_ids: Vec<_> = failures.iter().map(|(peer_id, _)| *peer_id).collect();
+
+                    let shards_disabled = tokio::task::spawn_blocking(move || {
+                        replica_state.wait_for(
+                            |state| {
+                                peer_ids.iter().all(|peer_id| {
+                                    state
+                                        .peers
+                                        .get(peer_id)
+                                        // Not found means that peer is dead
+                                        .map_or(true, |state| state != &ReplicaState::Active)
+                                })
+                            },
+                            timeout,
+                        )
+                    })
+                    .await?;
+
+                    if !shards_disabled {
+                        return Err(CollectionError::service_error(format!(
+                            "Some replica of shard {} failed to apply operation and deactivation \
+                            timed out after {} seconds. Consistency of this update is not guaranteed. Please retry. {failure_error}",
+                            self.shard_id, timeout.as_secs(),
+                        )));
+                    }
                 }
             }
         }
 
-        if !failures.is_empty() && successes.len() < minimal_success_count {
-            self.handle_failed_replicas(
-                failures
-                    .iter()
-                    .filter(|(peer_id, _)| self.peer_is_resharding(*peer_id)),
-                &self.replica_state.read(),
-                update_only_existing,
-            );
```

Comment:
Removing this is problematic. Keeping it seems to be problematic too.

Currently discussing what's best to do here with @ffuugoo. Will mark this as draft in the meanwhile.

Reply:
I reverted this bit to match the old behavior.

It may still not be ideal for resharding, but our plan is to handle it in a separate PR. Resharding is not released yet anyway.